Allow disabling of client type via allowed_clients param

chatton opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

This PR directly replaces #3194

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #2537


### Commit Message / Changelog Entry

N/A included in feature branch

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)


<!--
Example commit messages:

fix: skip emission of unpopulated memo field in ics20
deps: updating sdk to v0.46.4
chore: removed unused variables
e2e: adding e2e upgrade test for ibc-go/v6
docs: ics27 v6 documentation updates
feat: add semantic version utilities for e2e tests
feat(api)!: this is an api breaking feature
fix(statemachine)!: this is a statemachine breaking fix
-->

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

colin-axner commented on modules/core/02-client/keeper/keeper.go:
```diff
@@ -403,3 +403,12 @@ func (k Keeper) ClientStore(ctx sdk.Context, clientID string) sdk.KVStore {
 	return prefix.NewStore(ctx.KVStore(k.storeKey), clientPrefix)
 }
+
+// GetClientStatus returns the status for a given clientState. If the client type is not in the allowed
+// clients param field, Unauthorized is returned, otherwise the client state status is returned.
+func (k Keeper) GetClientStatus(ctx sdk.Context, clientState exported.ClientState, clientID string) exported.Status {
```
Should we add documentation somewhere that removing clients from the `AllowedClients` parameter will disable all usage of that client type?

chatton replied:
Yes I think we can improve this docstring a bit!

colin-axner reviewed (approved):
This is wonderful! So fast! We should probably add a changelog entry for this and I think it would be good to add a little docs on this functionality. It might be unclear when to use `clientState.Status()` vs `clientKeeper.GetClientStatus`

damiannolan reviewed (approved):
Beautiful, super clean PR! LGTM 🙌

chatton commented on docs/ibc/light-clients/client-state.md:
```diff
@@ -28,6 +28,8 @@ implementer is in charge of determining which checks are required. See the [Tend
 - A `Frozen` status indicates that misbehaviour was detected in the counterparty chain and the client is not allowed to be used.
 - An `Expired` status indicates that a client is not allowed to be used because it was not updated for longer than the trusting period.
 - An `Unknown` status indicates that there was an error in determining the status of a client.
+- An `Unauthorized` status indicates that usage of the client is not allowed.
+  - > Note: Removing a client type from the [`AllowedClients`](https://github.com/cosmos/ibc-go/blob/v6.0.0/modules/core/02-client/types/client.pb.go#L345) array will prevent all usage of that client type.
```
Added a note in the client-state.md doc. 

cc @colin-axner

colin-axner replied:
Hmm, the only issue with the doc here is that this is for the `clientState.Status` message. Light client developers don't need to use the `Unauthorized` status

chatton replied:
That is a very good point! Let me see if I can find a better place for this.

crodriguezvega commented on docs/ibc/light-clients/client-state.md:
```diff
@@ -28,6 +28,8 @@ implementer is in charge of determining which checks are required. See the [Tend
 - A `Frozen` status indicates that misbehaviour was detected in the counterparty chain and the client is not allowed to be used.
 - An `Expired` status indicates that a client is not allowed to be used because it was not updated for longer than the trusting period.
 - An `Unknown` status indicates that there was an error in determining the status of a client.
+- An `Unauthorized` status indicates that usage of the client is not allowed.
```
The formatting of the rendered text looks a bit weird with the `- >`. Was its use intentional?

chatton replied:
this was intentional to add a `note` section, but I think Colin brought up a good point that this is not the best place for these docs.

crodriguezvega reviewed (commented):
Nice work, @chatton!

Should we add some tests to exercise the new conditional checks added to `ChanOpenInit` and `ConnOpenInit`?

damiannolan commented:
Could we add the status check before updating the localhost client in `BeginBlocker` in this PR?

AdityaSripal commented on modules/core/04-channel/keeper/handshake.go:
```diff
@@ -50,6 +51,15 @@ func (k Keeper) ChanOpenInit(
+		return "", nil, sdkerrors.Wrapf(clienttypes.ErrClientNotFound, "clientID (%s)", connectionEnd.ClientId)
+	}
+
+	if status := k.clientKeeper.GetClientStatus(ctx, clientState, connectionEnd.ClientId); status != exported.Active {
```
Do we need this for ChanCloseInit as well?

AdityaSripal reviewed (commented):
I think we need a check on ChanCloseInit as well correct?

Also does the current test coverage cover the new k.ClientStatus function completely? Would be good to ensure we get an error if we try to call a method on a client that has a valid internal status but is not part of the AllowedClients params

chatton commented:
Based on PR feedback, I've 
* added a new Troubleshooting section to the docs
* added test cases which explicitly remove the tendermint client from the allow list
* added the check to `ChanCloseInit`
* added the check in `BeginBlocker`

AdityaSripal reviewed (approved):
Thanks for the quick fixes @chatton !

colin-axner reviewed (approved):
Excellent!
